pkg/ruler/manager/memstore.go

```diff
@@ -0,0 +1,380 @@
+	level.Debug(m.logger).Log("msg", "rule state successfully restored", "rule", ruleKey, "tenant", m.MemStore.userID, "len", len(vec))
+
+	// translate the result into the ALERTS_FOR_STATE series for caching,
+	// considered active & written at the timetamp requested
+	forStateVec := make(promql.Vector, 0, len(vec))
+	for _, smpl := range vec {
+
+		ts := util.TimeToMillis(m.ts)
+
+		forStateVec = append(forStateVec, promql.Sample{
+			Metric: ForStateMetric(smpl.Metric, rule.Name()),
+			Point: promql.Point{
+				T: ts,
+				V: float64(ts),
+			},
+		})
+
+	}
+
+	// cache the result of the evaluation at this timestamp
+	cache.Set(m.ts, forStateVec)
+
+	// Finally return the series if it exists
+	smpl, ok = cache.Get(m.ts, ls)
+	if !ok || smpl == nil {
+		return storage.NoopSeriesSet()
+	}
+	// If the labelset is cached we can consider it active. Return the for state sample active immediately.
+	return series.NewConcreteSeriesSet(
+		[]storage.Series{
+			series.NewConcreteSeries(smpl.Metric, []model.SamplePair{
+				{Timestamp: model.Time(util.TimeToMillis(m.ts)), Value: model.SampleValue(smpl.V)},
+			}),
+		},
+	)
+}
+
+// LabelValues returns all potential values for a label name.
+func (*memStoreQuerier) LabelValues(name string) ([]string, storage.Warnings, error) {
+	return nil, nil, errors.New("unimplemented")
+}
+
+// LabelNames returns all the unique label names present in the block in sorted order.
+func (*memStoreQuerier) LabelNames() ([]string, storage.Warnings, error) {
+	return nil, nil, errors.New("unimplemented")
+}
+
+// Close releases the resources of the Querier.
+func (*memStoreQuerier) Close() error { return nil }
+
+type RuleCache struct {
+	mtx     sync.Mutex
+	metrics *Metrics
+	data    map[int64]map[uint64]promql.Sample
+}
+
+func NewRuleCache(metrics *Metrics) *RuleCache {
+	return &RuleCache{
+		data:    make(map[int64]map[uint64]promql.Sample),
+		metrics: metrics,
+	}
+}
+
+func (c *RuleCache) Set(ts time.Time, vec promql.Vector) {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	tsMap, ok := c.data[ts.UnixNano()]
+	if !ok {
+		tsMap = make(map[uint64]promql.Sample)
+		c.data[ts.UnixNano()] = tsMap
+	}
+
+	for _, sample := range vec {
+		tsMap[sample.Metric.Hash()] = sample
+	}
+	c.metrics.Samples.Add(float64(len(vec)))
+}
+
+// Get returns ok if that timestamp's result is cached.
+func (c *RuleCache) Get(ts time.Time, ls labels.Labels) (*promql.Sample, bool) {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	match, ok := c.data[ts.UnixNano()]
+	if !ok {
+		return nil, false
+	}
+
+	smp, ok := match[ls.Hash()]
+	if !ok {
+		return nil, true
+	}
+	return &smp, true
+
+}
+
+// CleanupOldSamples removes samples that are outside of the rule's `For` duration.
+func (c *RuleCache) CleanupOldSamples(olderThan time.Time) (empty bool) {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	ns := olderThan.UnixNano()
+
+	// This could be more efficient (logarithmic instead of linear)
+	for ts, tsMap := range c.data {
+		if ts < ns {
+			delete(c.data, ts)
```

Comment:
I think this is dangerous? deleting from a slice while iterating over it? I think the pattern I usually see is to make a list of items to remove and then iterate that list to remove them?

Reply:
This should be fine (and is included in [effective go](https://golang.org/doc/effective_go.html#for))